src/core/pratt.ml

```diff
@@ -0,0 +1,120 @@
+(** A simple parser for operators. See Pratt parsing or the Shunting Yard
+    algorithm.
+
+    <https://dev.to/jrop/pratt-parsing>
+    <https://effbot.org/zone/simple-top-down-parsing.htm> *)
+
+open Lplib
+open Lplib.Extra
+
+open Syntax
+
+(** Table containing defined binary operators. *)
+let bin_operators : binop StrHtbl.t = StrHtbl.create 17
+
+(** Table containing defined unary operators. *)
+let una_operators : unop StrHtbl.t = StrHtbl.create 17
+
+module Pratt : sig
+  val expression : ?rbp:priority -> p_term Stream.t -> p_term
+end = struct
+  (** [lbp t] returns the binding power of term [t] (which is 0 if [t] is not
+      an operator). *)
+  let lbp : p_term -> priority = fun {elt=t; _} ->
+    match t with
+    | P_Iden({elt=(_,s); _}, _) ->
+        begin match StrHtbl.find_opt bin_operators s with
```

Comment:
Don't you assume here that s has already been scoped? Should be added in the comment.

Reply:
In fact, you seem to do scoping after applying Pratt, don't you? I think that you should scope idents before calling Pratt and use fully qualified names for looking in your table.

Reply:
It would require mutual dependencies between scoping and pratt parsing, if we want to traverse terms only once. Assuming we have term `(x s)`, we have to call `Pratt.parse` from scoping, then call scoping (of ident) from the Pratt parser. It may not be mutual dependent since we restrict to identifiers. But is there a clear advantage?

Reply:
Pratt parsing applies on sequences of terms. It only looks at the top of terms. In scoping, in case of an application, you could first do a pre-scoping of idents only, that is, if the term (head or argument) is a P_Iden s, call find_qid and replace s by its fully qualified name (if it is a symbol). This could be done by Pratt directly if you provide Pratt with (as additional argument) a function for knowing whether a P_Iden s is a symbol and, if so, what is its fully qualified name. To avoid calling find_qid twice, we could add a new_constructor P_Symb in p_term. This could be later improved in https://github.com/Deducteam/lambdapi/pull/500.

Reply:
I'm not sure I understand why you say that Pratt only looks at the _top_ of terms (or what do you mean by _top_?). But on the overall, I agree with your proposition to provide a more elaborate lookup function to Pratt.

Reply:
In pratt.ml there is no letrec.

Reply:
Ok ok I see. It doesn't recurse into wrapped terms nor binders is a way to put it formally.

Reply:
I'm not sure about which qid you're talking. Operators contain an identifier, the `+` in `set infix "+" := plus`, and a qid, the `plus`. Here we store the identifier, so that when we encounter a `P_iden("+")`, we are able to know that `"+"` is a defined binary operator. Which qid do you want to have when calling `find_qid`?

Reply:
Do you agree that a first problem is that your string s may be a variable equal to some operator name (e.g. \+,+*+)? To avoid this, you need to scope identifiers in their environment first.
Then, indeed, as operators is a map on unqualified identifiers, there is no need for having a qualified name. Btw, in case of a qualified name, you could improve the code as you don't even need to look in operators.
But what if the string is a qualified name? After all, if "+" is declared infix in some file A. Why not using A.+ as infix in some other file? This may be necessary if, in a single file, we want to use two distinct + at the same time, say A.+ and B.+. If you scope s first, then you will get its fully qualified name if it is a symbol. But, then, you need operators/pp_hints be maps on qualified names. But this point should probably be discussed in another PR.
Another question, that could be implemented in a separate PR too, is: why not replacing set infix/prefix commands by modifiers and, for instance, write `infix left symbol +:N->N->N`? Or do you think it is better to keep those infix/prefix commands?

Reply:
About "another question", we would like to delete `infix/prefix` commands but not now.
`infix left symbol +:N->N->N` is a good idea (Agda has a shorcut `infixl` for `infix left`), but we can do `infix left symbol _+_ : N->N->N` too.